### PR TITLE
Finalize Dev Branch

### PR DESCRIPTION
**Pull request by gusqgm**
Hi Tim-Oliver, here is the latest version of the dev branch of LSTree. I am still working on it, and need to check if everything merges nicely, but the structure will not change much from now on. I have mostly reorganized and added guides and comments to make the usability and understanding of the tools clearer.

Main points to check:

- [ ] Readability and clarity
- [ ] Ease to use example data run: from processing to webviewer experience

Main missing points / nice-to-haves:

- [ ] upload source data to *Zenodo* and link it here
- [ ] having a google colab to be able to run processing (and use the viewer) on the example data uploaded from *Zenodo* -> how to set it up?
- [ ] finalizing the addition of short explanation on the rest of the config file parameters.
- [ ] Adding a gif of the webviewer for better visualization of it

### CODE REVIEW

File: example/README.md [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+
+To train a model ( considerirng the existence of nuclei segmentation and lumen annotations ) run:
+
+```bash
+LUIGI_CONFIG_PATH=./config.cfg luigi --local-scheduler --module lstree CellSegmentationTrainingTask
+```
+
+Lumen segmentation and, if lineage tree exists, cell segmentation can be generated with:
```
Comment (tibuch) [inline example/README.md:99]: Why is this mentioned here?

File: example/README.md [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+Organoid-level features ( volume, lumen volume, etc ) and cell-level features ( cell/nuclei volume, aspect ratio, distance to lumen, etc ) can be extracted with:
+```bash
+LUIGI_CONFIG_PATH=./config.cfg luigi --local-scheduler --module lstree MultiAggregateFeaturesTask
+```
+or only organoid-level features if the lineage tree is not available with:
+```bash
+LUIGI_CONFIG_PATH=./config.cfg luigi --local-scheduler --module lstree MultiAggregateOrganoidFeaturesTask
+```
+
+It creates a *.csv file containing the extracted features for each timepoint as well as final a *.hdf file with all time-points aggregated. The *.hdf file contains Pandas DataFrames of nodes and edges with the associated features which can be used to plot lineage trees as illustrated in [plot_features.ipynb](notebooks/plot_features.ipynb).
```
Comment (tibuch) [inline example/README.md:114]: Dead link.

File: README.md [hunk 3]
```diff
@@ -119,152 +147,14 @@ max_patch_size=(9999,9999,9999)
...
-
----
-**NOTE - Generating training labels**
-
-
-**3D annotator**<br>
-<img src="docs/3D_annotator.png" width="800"/><br>
-
-The segmentation pipeline requires manual annotations of nuclei and lumen/epithelium. The provided [3D_annotator.ipynb](notebooks/3D_annotator.ipynb) can be used to generate 3D segmentation labels. Once an initial segmentation output is generated, it becomes much faster to handpick bad samples and correct them. The [segmentation_viewer.ipynb](notebooks/segmentation_viewer.ipynb) notebook allows quickly inspecting the current segmentation and copying files to be corrected in a separate folder in one click.
+More information is discussed within the [example guide](example/README.md).
```
Comment (tibuch) [inline README.md:150]: Where is the config discussed?

File: config_params.md [hunk 4]
```diff
@@ -0,0 +1,52 @@
+# Configuration file parameters
```
Comment (tibuch) [inline config_params.md:1]: What is the purpose of this file? How do users get here?

File: config.cfg [hunk 5]
```diff
@@ -1,11 +1,14 @@
...
 # General ##############################################################
 [DEFAULT]
 pattern={subdir}/{fname}-T{time:04d}.{ext}
 
-# set the same default number of threads of all multithreaded tasks
-# when processing a single dataset use n_threads ~= number of cores
-# Note that some tasks actually use multiprocessing when internal operations do not release the GIL
-#~n_threads=8
+### Important points:
+# 1) Set the same default number of threads of all multithreaded tasks
```
Comment (tibuch) [inline config.cfg:7]: How is that done?

File: config.cfg [hunk 6]
```diff
@@ -1,11 +1,14 @@
...
+# 2) When processing a single dataset use n_threads ~= number of cores
```
Comment (tibuch) [inline config.cfg:8]: What should be done if multiple datasets are processed?

File: config.cfg [hunk 7]
```diff
@@ -1,11 +1,14 @@
...
+# 3) Note that some tasks actually use multiprocessing when internal operations do not release the GIL.
```
Comment (tibuch) [inline config.cfg:9]: Which tasks? What does that mean for the config file? What is the GIL?

File: config.cfg [hunk 8]
```diff
@@ -1,11 +1,14 @@
...
+# 4) For all instances of 'movie_dirs' please provide a list of full paths to the corresponding channel folders containing the datasets, following the folder structure described in the main Readme of the Repository (https://github.com/fmi-basel/LSTree).
+# 5) Please n ote that all values given
```
Comment (tibuch) [inline config.cfg:11]: ?

File: config.cfg [hunk 9]
```diff
@@ -1,11 +1,14 @@
...
-# Note that some tasks actually use multiprocessing when internal operations do not release the GIL
-#~n_threads=8
+### Important points:
+# 1) Set the same default number of threads of all multithreaded tasks
+# 2) When processing a single dataset use n_threads ~= number of cores
+# 3) Note that some tasks actually use multiprocessing when internal operations do not release the GIL.
+# 4) For all instances of 'movie_dirs' please provide a list of full paths to the corresponding channel folders containing the datasets, following the folder structure described in the main Readme of the Repository (https://github.com/fmi-basel/LSTree).
+# 5) Please n ote that all values given 
 
 [resources]
```
Comment (tibuch) [inline config.cfg:13]: Maybe add some comments e.g. memory in MB or GB?

File: config.cfg [hunk 10]
```diff
@@ -15,7 +18,9 @@ memory=64000
 [core]
 workers=16
 log_level=INFO
+outdir = ./
 
+########################################################################
 # Denoising/Deconvolution ##############################################
 [BuildDenoiseTrainingRecordTask]
```
Comment (tibuch) [inline config.cfg:25]: What do these parameters mean?

File: config.cfg [hunk 11]
```diff
@@ -46,6 +51,8 @@ learning_rate=0.0001
 patch_size=(128,128,1)
 suffix=20200311
 
+# If a specific model needs to be retrained, or if a training stopped due to an error and needs to be continued, one can resume the weights based on "weights_latest.h5" or "weights_best.h5" from the previous model by uncommenting below.
+
 #resume_weights=PATH_TO_CURRENT_MODEL/weights_latest.h5
```
Comment (tibuch) [inline config.cfg:56]: Where is `PATH_TO_CURRENT_MODEL'? How can it be found?

File: config.cfg [hunk 12]
```diff
@@ -1,11 +1,14 @@
+########################################################################
```
Comment (tibuch) [inline config.cfg:1]: # General
There are many parameters and tasks in this file that are not explained. Seeing this for the first time is daunting.

File: lstree/config.py [hunk 13]
```diff
@@ -1,6 +1,6 @@
 import luigi
 
-SPACING = {25: (2, 0.26, 0.26), 37: (2, 0.173, 0.173)}
+SPACING = {25: (2, 0.26, 0.26), 37: (2, 0.173, 0.173)}  # check if this is explicitly used or if spacing comes solely from config.cfg and experiment.json files!
```
Comment (tibuch) [inline lstree/config.py:3]: Check?

File: notebooks/README.md [hunk 14]
```diff
@@ -0,0 +1,20 @@
+# Utilities
```
Comment (tibuch) [inline notebooks/README.md:1]: How are notebooks used?

File: webview/README.md [hunk 15]
```diff
@@ -0,0 +1,47 @@
...
+
+It is also possible to highlight certain areas of the tree by manual selection:  
+<img src="../docs/viewer_selection.png" width="1000"/><br>
+
+To be responsive, the viewer relies on smoothed 3D meshes pre-generated from the segmentation stacks with:
+```bash
+LUIGI_CONFIG_PATH=./config.cfg luigi --local-scheduler --module lstree ViewerTask
+```
+
+## Running via the notebook / port forwarding
```
Comment (tibuch) [inline webview/README.md:22]: Would it make sense to move this section to the top?

**Review by tibuch — COMMENTED** [review]
I looked at the markdown files and added my remarks/suggestions. 

# General
In my opinion a visualization showing what LSTree is and how the data flows would be helpful in building understanding. Such a visualization could also distinguish between training all the models and applying LSTree after training.

# Config
Every parameter and task in the config file requires a description.

# Test installation and run
I did _not_ install LSTree and I did _not_ run any of the code. Did anyone do this from scratch and verified that everything works?

File: example/README.md [hunk 16]
```diff
@@ -0,0 +1,145 @@
...
+```
+
+## 3. Lineage tree
+LSTree works directly directly with exported MaMuT `.xml`git files that contain trees made using [Mastodon](https://github.com/mastodon-sc/mastodon) Fiji plugin. Subsequently a deep learning model can be (re)trained to predict trees that require fewer manual corrections. Alternatively, one can also use the output from [Elephant](https://elephant-track.github.io/#/v0.3/) ( as MaMuT `.xml` ) to fine tune an existing model or just as final lineage tree output for further processing (segmentation and feature extraction steps). 
+
+```bash
+LUIGI_CONFIG_PATH=./config.cfg luigi --local-scheduler --module lstree MultiBuildTreeTask
+```
+
+The tree prediction outputs include a `.xml` file that can be opened with Mastodon as well as a plots of the predicted tree with overlaid information to facilitate corrections (e.g. nuclei volume, tracking embedding distance).
```
[the remark this answers is not in the document]
Comment (gusqgm) [reply example/README.md:60]: this is part of the lineage tree prediction

File: webview/README.md [hunk 17]
```diff
@@ -0,0 +1,51 @@
...
+# Introduction - Webviewer
+
+The included web-based viewer allows visualizing a lineage tree with a linked view of the 3D cell/nuclei segmentation at a given timepoint.  
+
+<!-- ![webviewer](../docs/webviewer_presentation.mp4) -->
+
+## Running via the notebook / port forwarding
+You can run the Webviewer via the jupyter [notebook](webview.ipynb). This is a good first way to look at the tree-meshes visualization, as it is set to look initially at the processed example datasets.
+
+Alternatively, you can also use port forwarding to serve trhe notebook for remote visualization.
```
Comment (imagejan) [inline webview/README.md:11]:
```suggestion
Alternatively, you can also use port forwarding to serve the notebook for remote visualization.
```

File: webview/README.md [hunk 18]
```diff
@@ -0,0 +1,51 @@
...
+Here, `PATH_TO_PROCESSED_DATA` is the full path to where the repository or the datasets folders ( following the structure presented in the [Usage](../README.md) section ).
+
+or on a remote machine:
+
+```bash
+cd /PATH_TO_LSTREE_REPOSITORY/webview
+panel serve --static-dirs static_data="static_data" --port PORT --allow-websocket-origin=WORKSTATION:PORT webview.ipynb  --args --basedir PATH_TO_PROCESSED_DATA
+```
+
+Example, considering that the data is located in /Data:
```
Comment (imagejan) [inline webview/README.md:29]:
```suggestion
Example, considering that the data is located in `/Data`:
```

or maybe the full path `/Data/LSTree/example/data` to be clear?

**Review by imagejan — COMMENTED** [review]
Just two small things I spotted when reading the webviewer docs.
